ledger/participant-integration-api/src/test/suite/scala/platform/indexer/parallel/BatchingParallelIngestionPipeSpec.scala

```diff
@@ -0,0 +1,169 @@
+
+import akka.NotUsed
+import akka.stream.scaladsl.Source
+import com.daml.ledger.api.testing.utils.AkkaBeforeAndAfterAll
+import org.scalatest.flatspec.AsyncFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.concurrent.duration.FiniteDuration
+
+class BatchingParallelIngestionPipeSpec
+    extends AsyncFlatSpec
+    with Matchers
+    with AkkaBeforeAndAfterAll {
+
+  // AsyncFlatSpec is with serial execution context
+  private implicit val ec: ExecutionContext = system.dispatcher
+
+  private val input = Iterator.continually(util.Random.nextInt()).take(100).toList
+
+  it should "end the stream successfully in a happy path case" in {
+    runPipe().map { case (ingested, ingestedTail, err) =>
+      err shouldBe empty
+      ingested.sortBy(_._1) shouldBe input.map(_.toString).zipWithIndex.map { case (s, i) =>
+        (i + 1, s)
+      }
+      ingestedTail.last shouldBe 100
+    }
+  }
+
+  it should "terminate the stream upon error in input mapper" in {
+    runPipe(inputMapperHook = () => throw new Exception("inputmapper failed")).map {
+      case (_, _, err) =>
+        err should not be empty
+        err.get.getMessage shouldBe "inputmapper failed"
+    }
+  }
+
+  it should "terminate the stream upon error in seqMapper" in {
+    runPipe(seqMapperHook = () => throw new Exception("seqMapper failed")).map { case (_, _, err) =>
+      err should not be empty
+      err.get.getMessage shouldBe "seqMapper failed"
+    }
+  }
+
+  it should "terminate the stream upon error in batcher" in {
+    runPipe(batcherHook = () => throw new Exception("batcher failed")).map { case (_, _, err) =>
+      err should not be empty
+      err.get.getMessage shouldBe "batcher failed"
+    }
+  }
+
+  it should "terminate the stream upon error in ingester" in {
+    runPipe(ingesterHook = _ => throw new Exception("ingester failed")).map { case (_, _, err) =>
+      err should not be empty
+      err.get.getMessage shouldBe "ingester failed"
+    }
+  }
+
+  it should "terminate the stream upon error in tailer" in {
+    runPipe(tailerHook = () => throw new Exception("tailer failed")).map { case (_, _, err) =>
+      err should not be empty
+      err.get.getMessage shouldBe "tailer failed"
+    }
+  }
+
+  it should "terminate the stream upon error in ingestTail" in {
+    runPipe(ingestTailHook = () => throw new Exception("ingestTail failed")).map {
+      case (_, _, err) =>
+        err should not be empty
+        err.get.getMessage shouldBe "ingestTail failed"
+    }
+  }
+
+  it should "hold the stream if a single ingestion takes too long" in {
+    runPipe(
+      ingesterHook = batch => {
+        if (batch.head == 21) Thread.sleep(1000)
+      },
+      timeout = FiniteDuration(100, "milliseconds"),
+    ).map { case (ingested, ingestedTail, err) =>
+      err should not be empty
+      err.get.getMessage shouldBe "timed out"
+      ingested.size shouldBe 25
```

Comment:
Why is this 25?

The test times out the 21st "batch", but the assertion is done on the number of "state updates" ingested. Does this test depend on the timing how updates are batched together?

Reply:
no, as soon as the 21..25 batch is blocked, we still process the 26..30 batch, and only then the mapAsync stage is forced to block as well